File: base/src/main/vsdk/toolkit/common/statistics/PolyhedralBoundedSolidStatistics.java
package vsdk.toolkit.common;

import java.util.concurrent.atomic.LongAdder;

public final class PolyhedralBoundedSolidStatistics {
    private static final boolean ENABLED =
        Boolean.getBoolean("vsdk.polyhedral.stats");

    private static final LongAdder eulerLmevCalls = new LongAdder();
    private static final LongAdder eulerLkevCalls = new LongAdder();
    private static final LongAdder eulerLkefCalls = new LongAdder();
    private static final LongAdder eulerLmefCalls = new LongAdder();
    private static final LongAdder eulerLkemrCalls = new LongAdder();
    private static final LongAdder eulerLmekrCalls = new LongAdder();
    private static final LongAdder eulerLringmvCalls = new LongAdder();

    private static final LongAdder setOpCalls = new LongAdder();
    private static final LongAdder setOpUnionCalls = new LongAdder();
    private static final LongAdder setOpIntersectionCalls = new LongAdder();
    private static final LongAdder setOpSubtractCalls = new LongAdder();

    private static final LongAdder splitCalls = new LongAdder();
    private static final LongAdder splitNoNullEdgesCases = new LongAdder();
    private static final LongAdder splitProducedAboveSolids = new LongAdder();
    private static final LongAdder splitProducedBelowSolids = new LongAdder();

    private static final LongAdder joinCalls = new LongAdder();
    private static final LongAdder joinIncompleteCases = new LongAdder();

    private static final LongAdder he1EqualsHe2Cases = new LongAdder();
    private static final LongAdder invalidHalfEdgeInputCases = new LongAdder();
    private static final LongAdder consistencyWarningCases = new LongAdder();
    private static final LongAdder operationFailureCases = new LongAdder();

    private PolyhedralBoundedSolidStatistics() {
    }

    public static boolean isEnabled() {
        return ENABLED;
    }

    public static void recordLmevCall() {
        if (ENABLED) {
            eulerLmevCalls.increment();
        }
    }

    public static void recordLkevCall() {
        if (ENABLED) {
            eulerLkevCalls.increment();
        }
    }

    public static void recordLkefCall() {
        if (ENABLED) {
            eulerLkefCalls.increment();
        }
    }

    public static void recordLmefCall() {
        if (ENABLED) {
            eulerLmefCalls.increment();
        }
    }

    public static void recordLkemrCall() {
        if (ENABLED) {
            eulerLkemrCalls.increment();
        }
    }

    public static void recordLmekrCall() {
        if (ENABLED) {
            eulerLmekrCalls.increment();
        }
    }

    public static void recordLringmvCall() {
        if (ENABLED) {
            eulerLringmvCalls.increment();
        }
    }

    public static void recordSetOpCall(int op) {
        if (!ENABLED) {
            return;
        }
        setOpCalls.increment();
        if (op == 1) {
            setOpUnionCalls.increment();
        }
        else if (op == 2) {
            setOpIntersectionCalls.increment();
        }
        else if (op == 3) {
            setOpSubtractCalls.increment();
        }
    }

    public static void recordSplitCall() {
        if (ENABLED) {
            splitCalls.increment();
        }
    }

    public static void recordSplitNoNullEdgesCase() {
        if (ENABLED) {
            splitNoNullEdgesCases.increment();
        }
    }

    public static void recordSplitProducedSolids(int above, int below) {
        if (!ENABLED) {
            return;
        }
        if (above > 0) {
            splitProducedAboveSolids.add(above);
        }
        if (below > 0) {
            splitProducedBelowSolids.add(below);
        }
    }

    public static void recordJoinCall() {
        if (ENABLED) {
            joinCalls.increment();
        }
    }

    public static void recordJoinIncompleteCase() {
        if (ENABLED) {
            joinIncompleteCases.increment();
        }
    }

    public static void recordHe1EqualsHe2Case() {
        if (ENABLED) {
            he1EqualsHe2Cases.increment();
        }
    }

    public static void recordInvalidHalfEdgeInputCase() {
        if (ENABLED) {
            invalidHalfEdgeInputCases.increment();
        }
    }

    public static void recordConsistencyWarningCase() {
        if (ENABLED) {
            consistencyWarningCases.increment();
        }
    }

    public static void recordOperationFailureCase() {
        if (ENABLED) {
            operationFailureCases.increment();
        }
    }

    public static void reset() {
        if (!ENABLED) {
            return;
        }
        eulerLmevCalls.reset();
        eulerLkevCalls.reset();
        eulerLkefCalls.reset();
        eulerLmefCalls.reset();
        eulerLkemrCalls.reset();
        eulerLmekrCalls.reset();
        eulerLringmvCalls.reset();

        setOpCalls.reset();
        setOpUnionCalls.reset();
        setOpIntersectionCalls.reset();
        setOpSubtractCalls.reset();

        splitCalls.reset();
        splitNoNullEdgesCases.reset();
        splitProducedAboveSolids.reset();
        splitProducedBelowSolids.reset();

        joinCalls.reset();
        joinIncompleteCases.reset();

        he1EqualsHe2Cases.reset();
        invalidHalfEdgeInputCases.reset();
        consistencyWarningCases.reset();
        operationFailureCases.reset();
    }

    public static long getOperationFailureCases() {
        return operationFailureCases.sum();
    }

    public static long getConsistencyWarningCases() {
        return consistencyWarningCases.sum();
    }

    public static long getHe1EqualsHe2Cases() {
        return he1EqualsHe2Cases.sum();
    }

    public static long getInvalidHalfEdgeInputCases() {
        return invalidHalfEdgeInputCases.sum();
    }

    public static long getJoinIncompleteCases() {
        return joinIncompleteCases.sum();
    }

    public static long getSetOpCalls() {
        return setOpCalls.sum();
    }

    public static void printSummary() {
        long lmev = eulerLmevCalls.sum();
        long lkev = eulerLkevCalls.sum();
        long lkef = eulerLkefCalls.sum();
        long lmef = eulerLmefCalls.sum();
        long lkemr = eulerLkemrCalls.sum();
        long lmekr = eulerLmekrCalls.sum();
        long lringmv = eulerLringmvCalls.sum();
        long eulerTotal = lmev + lkev + lkef + lmef + lkemr + lmekr + lringmv;

        System.out.println("PolyhedralBoundedSolid statistics (enabled=" +
            ENABLED + "):");
        System.out.println("  Euler ops total: " + eulerTotal);
        System.out.println("    lmev: " + lmev);
        System.out.println("    lkev: " + lkev);
        System.out.println("    lkef: " + lkef);
        System.out.println("    lmef: " + lmef);
        System.out.println("    lkemr: " + lkemr);
        System.out.println("    lmekr: " + lmekr);
        System.out.println("    lringmv: " + lringmv);

        System.out.println("  Boolean setOp calls: " + setOpCalls.sum());
        System.out.println("    union: " + setOpUnionCalls.sum());
        System.out.println("    intersection: " + setOpIntersectionCalls.sum());
        System.out.println("    subtract: " + setOpSubtractCalls.sum());

        System.out.println("  Slicing split calls: " + splitCalls.sum());
        System.out.println("    no-null-edges fast-path: " + splitNoNullEdgesCases.sum());
        System.out.println("    produced above solids: " + splitProducedAboveSolids.sum());
        System.out.println("    produced below solids: " + splitProducedBelowSolids.sum());

        System.out.println("  Join calls: " + joinCalls.sum());
        System.out.println("  Join incomplete cases: " + joinIncompleteCases.sum());

        System.out.println("  Borderline he1 == he2 cases: " + he1EqualsHe2Cases.sum());
        System.out.println("  Invalid half-edge inputs: " + invalidHalfEdgeInputCases.sum());
        System.out.println("  Consistency warnings: " + consistencyWarningCases.sum());
        System.out.println("  Operation failure cases: " + operationFailureCases.sum());
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/statistics/PolyhedralBoundedSolidStatistics.ts
type StatisticsFlags = { readonly raytrace?: boolean; readonly polyhedral?: boolean };
const statisticsFlags = (globalThis as { __vitralStatistics?: StatisticsFlags }).__vitralStatistics;

/** Opt-in browser-safe counterpart of the Java system-property polyhedral counters. */
export class PolyhedralBoundedSolidStatistics {
    private static readonly enabled = statisticsFlags?.polyhedral === true;
    private static lmev = 0n;
    private static lkev = 0n;
    private static lkef = 0n;
    private static lmef = 0n;
    private static lkemr = 0n;
    private static lmekr = 0n;
    private static lringmv = 0n;
    private static setOperations = 0n;
    private static unions = 0n;
    private static intersections = 0n;
    private static subtracts = 0n;
    private static splits = 0n;
    private static splitNoNullEdges = 0n;
    private static splitAbove = 0n;
    private static splitBelow = 0n;
    private static joins = 0n;
    private static incompleteJoins = 0n;
    private static equalHalfEdges = 0n;
    private static invalidHalfEdges = 0n;
    private static consistencyWarnings = 0n;
    private static operationFailures = 0n;
    public static isEnabled(): boolean {
        return this.enabled;
    }
    public static recordLmevCall(): void {
        if (this.enabled) this.lmev++;
    }
    public static recordLkevCall(): void {
        if (this.enabled) this.lkev++;
    }
    public static recordLkefCall(): void {
        if (this.enabled) this.lkef++;
    }
    public static recordLmefCall(): void {
        if (this.enabled) this.lmef++;
    }
    public static recordLkemrCall(): void {
        if (this.enabled) this.lkemr++;
    }
    public static recordLmekrCall(): void {
        if (this.enabled) this.lmekr++;
    }
    public static recordLringmvCall(): void {
        if (this.enabled) this.lringmv++;
    }
    public static recordSetOpCall(operation: number): void {
        if (!this.enabled) return;
        this.setOperations++;
        if (operation === 1) this.unions++;
        else if (operation === 2) this.intersections++;
        else if (operation === 3) this.subtracts++;
    }
    public static recordSplitCall(): void {
        if (this.enabled) this.splits++;
    }
    public static recordSplitNoNullEdgesCase(): void {
        if (this.enabled) this.splitNoNullEdges++;
    }
    public static recordSplitProducedSolids(above: number, below: number): void {
        if (!this.enabled) return;
        if (above > 0) this.splitAbove += BigInt(above);
        if (below > 0) this.splitBelow += BigInt(below);
    }
    public static recordJoinCall(): void {
        if (this.enabled) this.joins++;
    }
    public static recordJoinIncompleteCase(): void {
        if (this.enabled) this.incompleteJoins++;
    }
    public static recordHe1EqualsHe2Case(): void {
        if (this.enabled) this.equalHalfEdges++;
    }
    public static recordInvalidHalfEdgeInputCase(): void {
        if (this.enabled) this.invalidHalfEdges++;
    }
    public static recordConsistencyWarningCase(): void {
        if (this.enabled) this.consistencyWarnings++;
    }
    public static recordOperationFailureCase(): void {
        if (this.enabled) this.operationFailures++;
    }
    public static reset(): void {
        if (!this.enabled) return;
        this.lmev =
            this.lkev =
            this.lkef =
            this.lmef =
            this.lkemr =
            this.lmekr =
            this.lringmv =
            this.setOperations =
            this.unions =
            this.intersections =
            this.subtracts =
            this.splits =
            this.splitNoNullEdges =
            this.splitAbove =
            this.splitBelow =
            this.joins =
            this.incompleteJoins =
            this.equalHalfEdges =
            this.invalidHalfEdges =
            this.consistencyWarnings =
            this.operationFailures =
                0n;
    }
    public static getOperationFailureCases(): bigint {
        return this.operationFailures;
    }
    public static getConsistencyWarningCases(): bigint {
        return this.consistencyWarnings;
    }
    public static getHe1EqualsHe2Cases(): bigint {
        return this.equalHalfEdges;
    }
    public static getInvalidHalfEdgeInputCases(): bigint {
        return this.invalidHalfEdges;
    }
    public static getJoinIncompleteCases(): bigint {
        return this.incompleteJoins;
    }
    public static getSetOpCalls(): bigint {
        return this.setOperations;
    }
    public static printSummary(): void {
        if (!this.enabled) return;
        const eulerTotal = this.lmev + this.lkev + this.lkef + this.lmef + this.lkemr + this.lmekr + this.lringmv;
        console.log(
            `PolyhedralBoundedSolid statistics (enabled=true):\n  Euler ops total: ${eulerTotal}\n    lmev: ${this.lmev}\n    lkev: ${this.lkev}\n    lkef: ${this.lkef}\n    lmef: ${this.lmef}\n    lkemr: ${this.lkemr}\n    lmekr: ${this.lmekr}\n    lringmv: ${this.lringmv}\n  Boolean setOp calls: ${this.setOperations}\n    union: ${this.unions}\n    intersection: ${this.intersections}\n    subtract: ${this.subtracts}\n  Slicing split calls: ${this.splits}\n    no-null-edges fast-path: ${this.splitNoNullEdges}\n    produced above solids: ${this.splitAbove}\n    produced below solids: ${this.splitBelow}\n  Join calls: ${this.joins}\n  Join incomplete cases: ${this.incompleteJoins}\n  Borderline he1 == he2 cases: ${this.equalHalfEdges}\n  Invalid half-edge inputs: ${this.invalidHalfEdges}\n  Consistency warnings: ${this.consistencyWarnings}\n  Operation failure cases: ${this.operationFailures}`,
        );
    }
}
